### PR TITLE
Fix: warmup_bias_lr override for MuSGD in auto optimizer mode

### PR DESCRIPTION
**Pull request by songjiahao-wq**
Fix warmup_bias_lr override for MuSGD in auto optimizer mode

  Description:

  Summary

  - Move warmup_bias_lr=0.0 override to only apply when AdamW is selected, not MuSGD
  - Log the auto-selected optimizer parameters after override for better debuggability

  Changes

  ultralytics/engine/trainer.py:

```python
  # Before (line 951-952):
  name, lr, momentum = ("MuSGD", 0.01, 0.9) if iterations > 10000 else ("AdamW", lr_fit, 0.9)
  self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam
```

```python
  # After:
  if iterations > 10000:
      name, lr, momentum = "MuSGD", 0.01, 0.9
  else:
      name, lr, momentum = "AdamW", lr_fit, 0.9
      self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam
```

  Problem

When optimizer=auto selects MuSGD (iterations > 10000), warmup_bias_lr is incorrectly forced to 0.0. This setting was intended for Adam-family optimizers only (as noted in the `comment: "no higher than 0.01 for Adam"), but applies unconditionally.`)
 
Additionally, print_args() is called before build_optimizer(), so the log always shows the original warmup_bias_lr=0.1 even when it has been overridden to 0.0, making this issue invisible to users.

```
  Test

  Same dataset (1795 images), same config (400 epochs, seed=0, MuSGD, lr=0.01, momentum=0.9):
  ┌─────────────────────────────┬───────────────────────────┬─────────────────┐
  │           Setting           │  warmup_bias_lr (actual)  │     Result      │
  ├─────────────────────────────┼───────────────────────────┼─────────────────┤
  │ optimizer=auto (before fix) │ 0.0 (silently overridden) │ Lower accuracy  │
  ├─────────────────────────────┼───────────────────────────┼─────────────────┤
  │ optimizer=MuSGD (manual)    │ 0.1 (default)             │ Better accuracy │
  └─────────────────────────────┴───────────────────────────┴─────────────────┘
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fixes an unintended `warmup_bias_lr` override when auto-selecting MuSGD vs AdamW during optimizer auto-configuration. 🛠️

### 📊 Key Changes
- Restricts `self.args.warmup_bias_lr = 0.0` to only the AdamW auto-optimizer path
- Refactors the auto-optimizer selection into an explicit `if/else` block for clarity
- Preserves MuSGD behavior for long-training runs without silently changing warmup bias LR

### 🎯 Purpose & Impact
- Prevents MuSGD runs from inheriting AdamW-specific warmup bias LR settings ✅
- Improves correctness and predictability of training hyperparameters in `optimizer=auto` mode 📈
- Reduces confusion when switching optimizers based on dataset iterations, especially for large datasets 🔍

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @songjiahao-wq, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message, and an Ultralytics engineer will assist soon. To ensure a seamless integration of your work, please review the following checklist:

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Looks good to me. The refactor cleanly limits warmup_bias_lr changes to the AdamW path and preserves MuSGD behavior. No issues found.

**Comment by songjiahao-wq** [comment]
I have read the CLA Document and I sign the CLA

**Comment by songjiahao-wq** [comment]
it's ready for merge.

**Comment by glenn-jocher** [comment]
Thanks — the change makes sense; we’ll review and merge once CI is green and the review is complete (no ETA), and if you can add a small test that confirms `optimizer=auto` only forces `warmup_bias_lr=0.0` for AdamW (not when it selects `MuSGD`), that would help per our [contributing guide](https://docs.ultralytics.com/help/contributing/).

**Comment by songjiahao-wq** [comment]
```
# Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license

import sys
from unittest import mock

import torch

from tests import MODEL, SOURCE
from ultralytics import YOLO
from ultralytics.cfg import get_cfg
from ultralytics.engine.exporter import Exporter
from ultralytics.models.yolo import classify, detect, segment
from ultralytics.utils import ASSETS, DEFAULT_CFG, WEIGHTS_DIR


def test_func(*args, **kwargs):
    """Test function callback for evaluating YOLO model performance metrics."""
    print("callback test passed")


def test_export():
    """Test model exporting functionality by adding a callback and verifying its execution."""
    exporter = Exporter()
    exporter.add_callback("on_export_start", test_func)
    assert test_func in exporter.callbacks["on_export_start"], "callback test failed"
    f = exporter(model=YOLO("yolo26n.yaml").model)
    YOLO(f)(SOURCE)  # exported model inference


def test_detect():
    """Test YOLO object detection training, validation, and prediction functionality."""
    overrides = {"data": "coco8.yaml", "model": "yolo26n.yaml", "imgsz": 32, "epochs": 1, "save": False}
    cfg = get_cfg(DEFAULT_CFG)
    cfg.data = "coco8.yaml"
    cfg.imgsz = 32

    # Trainer
    trainer = detect.DetectionTrainer(overrides=overrides)
    trainer.add_callback("on_train_start", test_func)
    assert test_func in trainer.callbacks["on_train_start"], "callback test failed"
    trainer.train()

    # Validator
    val = detect.DetectionValidator(args=cfg)
    val.add_callback("on_val_start", test_func)
    assert test_func in val.callbacks["on_val_start"], "callback test failed"
    val(model=trainer.best)  # validate best.pt

    # Predictor
    pred = detect.DetectionPredictor(overrides={"imgsz": [64, 64]})
    pred.add_callback("on_predict_start", test_func)
    assert test_func in pred.callbacks["on_predict_start"], "callback test failed"
    # Confirm there is no issue with sys.argv being empty
    with mock.patch.object(sys, "argv", []):
        result = pred(source=ASSETS, model=MODEL)
        assert len(result), "predictor test failed"

    # Test resume functionality
    overrides["resume"] = trainer.last
    trainer = detect.DetectionTrainer(overrides=overrides)
    try:
        trainer.train()
    except Exception as e:
        print(f"Expected exception caught: {e}")
        return

    raise Exception("Resume test failed!")


def test_segment():
    """Test image segmentation training, validation, and prediction pipelines using YOLO models."""
    overrides = {
        "data": "coco8-seg.yaml",
        "model": "yolo26n-seg.yaml",
        "imgsz": 32,
        "epochs": 1,
        "save": False,
        "mask_ratio": 1,
        "overlap_mask": False,
    }
    cfg = get_cfg(DEFAULT_CFG)
    cfg.data = "coco8-seg.yaml"
    cfg.imgsz = 32

    # Trainer
    trainer = segment.SegmentationTrainer(overrides=overrides)
    trainer.add_callback("on_train_start", test_func)
    assert test_func in trainer.callbacks["on_train_start"], "callback test failed"
    trainer.train()

    # Validator
    val = segment.SegmentationValidator(args=cfg)
    val.add_callback("on_val_start", test_func)
    assert test_func in val.callbacks["on_val_start"], "callback test failed"
    val(model=trainer.best)  # validate best.pt

    # Predictor
    pred = segment.SegmentationPredictor(overrides={"imgsz": [64, 64]})
    pred.add_callback("on_predict_start", test_func)
    assert test_func in pred.callbacks["on_predict_start"], "callback test failed"
    result = pred(source=ASSETS, model=WEIGHTS_DIR / "yolo26n-seg.pt")
    assert len(result), "predictor test failed"

    # Test resume functionality
    overrides["resume"] = trainer.last
    trainer = segment.SegmentationTrainer(overrides=overrides)
    try:
        trainer.train()
    except Exception as e:
        print(f"Expected exception caught: {e}")
        return

    raise Exception("Resume test failed!")


def test_auto_optimizer_warmup_bias_lr():
    """Test that warmup_bias_lr is only forced to 0.0 for AdamW, not MuSGD, in auto optimizer mode."""
    overrides = {"data": "coco8.yaml", "model": "yolo26n.yaml", "imgsz": 32, "epochs": 1, "save": False}
    trainer = detect.DetectionTrainer(overrides=overrides)
    model = YOLO("yolo26n.yaml").model

    # Case 1: iterations > 10000 → auto selects MuSGD → warmup_bias_lr should NOT be overridden
    original_warmup_bias_lr = trainer.args.warmup_bias_lr
    trainer.build_optimizer(model, name="auto", iterations=20000)
    assert trainer.args.warmup_bias_lr == original_warmup_bias_lr, (
        f"warmup_bias_lr should not change for MuSGD, "
        f"expected {original_warmup_bias_lr}, got {trainer.args.warmup_bias_lr}"
    )

    # Case 2: iterations <= 10000 → auto selects AdamW → warmup_bias_lr should be forced to 0.0
    trainer.args.warmup_bias_lr = 0.1  # reset to non-zero value
    trainer.build_optimizer(model, name="auto", iterations=5000)
    assert trainer.args.warmup_bias_lr == 0.0, (
        f"warmup_bias_lr should be 0.0 for AdamW, got {trainer.args.warmup_bias_lr}"
    )
    """Test image classification including training, validation, and prediction phases."""
    overrides = {"data": "imagenet10", "model": "yolo26n-cls.yaml", "imgsz": 32, "epochs": 1, "save": False}
    cfg = get_cfg(DEFAULT_CFG)
    cfg.data = "imagenet10"
    cfg.imgsz = 32

    # Trainer
    trainer = classify.ClassificationTrainer(overrides=overrides)
    trainer.add_callback("on_train_start", test_func)
    assert test_func in trainer.callbacks["on_train_start"], "callback test failed"
    trainer.train()

    # Validator
    val = classify.ClassificationValidator(args=cfg)
    val.add_callback("on_val_start", test_func)
    assert test_func in val.callbacks["on_val_start"], "callback test failed"
    val(model=trainer.best)

    # Predictor
    pred = classify.ClassificationPredictor(overrides={"imgsz": [64, 64]})
    pred.add_callback("on_predict_start", test_func)
    assert test_func in pred.callbacks["on_predict_start"], "callback test failed"
    result = pred(source=ASSETS, model=trainer.best)
    assert len(result), "predictor test failed"


def test_nan_recovery():
    """Test NaN loss detection and recovery during training."""
    nan_injected = [False]

    def inject_nan(trainer):
        """Inject NaN into loss during batch processing to test recovery mechanism."""
        if trainer.epoch == 1 and trainer.tloss is not None and not nan_injected[0]:
            trainer.tloss *= torch.tensor(float("nan"))
            nan_injected[0] = True

    overrides = {"data": "coco8.yaml", "model": "yolo26n.yaml", "imgsz": 32, "epochs": 3}
    trainer = detect.DetectionTrainer(overrides=overrides)
    trainer.add_callback("on_train_batch_end", inject_nan)
    trainer.train()
    assert nan_injected[0], "NaN injection failed"

```
I modified tests/test_engine.py for the test purposes. Below are the parameters printed.
image

**Comment by glenn-jocher** [comment]
Thanks for adding a test — could you trim this down to a small, fast unit test that only covers the `optimizer=auto` side-effect in `BaseTrainer.build_optimizer()` (i.e. `warmup_bias_lr` forced to `0.0` only when auto selects `AdamW`, not when it selects `MuSGD`), and avoid bundling export/train/val/predict/resume coverage into the same change? The intended behavior is in the [trainer `build_optimizer()` reference](https://docs.ultralytics.com/reference/engine/trainer/).

```python
def test_auto_optimizer_warmup_bias_lr():
    from ultralytics import YOLO
    from ultralytics.models.yolo.detect import DetectionTrainer

    trainer = DetectionTrainer(overrides={"data": "coco8.yaml", "model": "yolo26n.yaml", "imgsz": 32, "epochs": 1, "save": False})
    model = YOLO("yolo26n.yaml").model

    trainer.args.warmup_bias_lr = 0.1
    opt = trainer.build_optimizer(model, name="auto", iterations=20001)
    assert type(opt).__name__ == "MuSGD"
    assert trainer.args.warmup_bias_lr == 0.1

    trainer.args.warmup_bias_lr = 0.1
    opt = trainer.build_optimizer(model, name="auto", iterations=1)
    assert type(opt).__name__ == "AdamW"
    assert trainer.args.warmup_bias_lr == 0.0
```

**Comment by songjiahao-wq** [comment]
```
from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer


def test_auto_optimizer_warmup_bias_lr():
    """Test that warmup_bias_lr is only forced to 0.0 for AdamW, not MuSGD, in auto optimizer mode."""
    trainer = DetectionTrainer(overrides={"data": "coco8.yaml", "model": "yolo26n.yaml", "imgsz": 32, "epochs": 1, "save": False})
    model = YOLO("yolo26n.yaml").model

    trainer.args.warmup_bias_lr = 0.1
    opt = trainer.build_optimizer(model, name="auto", iterations=20001)
    assert type(opt).__name__ == "MuSGD"
    assert trainer.args.warmup_bias_lr == 0.1

    trainer.args.warmup_bias_lr = 0.1
    opt = trainer.build_optimizer(model, name="auto", iterations=1)
    assert type(opt).__name__ == "AdamW"
    assert trainer.args.warmup_bias_lr == 0.0


if __name__ == "__main__":
    test_auto_optimizer_warmup_bias_lr()

```
image
My output result is correct.

**Comment by glenn-jocher** [comment]
That looks good — please commit it as a `pytest` unit test (remove the `if __name__ == "__main__":` block and the screenshot) in `tests/test_engine.py` (or a small new test file), and confirm it runs cleanly with:

```bash
pytest -k test_auto_optimizer_warmup_bias_lr
```

Once CI is green we’ll continue with the review/merge process (no ETA).

**Comment by songjiahao-wq** [comment]
```
from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer


def test_auto_optimizer_warmup_bias_lr():
    """Test that warmup_bias_lr is only forced to 0.0 for AdamW, not MuSGD, in auto optimizer mode."""
    trainer = DetectionTrainer(overrides={"data": "coco8.yaml", "model": "yolo26n.yaml", "imgsz": 32, "epochs": 1, "save": False})
    model = YOLO("yolo26n.yaml").model

    trainer.args.warmup_bias_lr = 0.1
    opt = trainer.build_optimizer(model, name="auto", iterations=20001)
    assert type(opt).__name__ == "MuSGD"
    assert trainer.args.warmup_bias_lr == 0.1

    trainer.args.warmup_bias_lr = 0.1
    opt = trainer.build_optimizer(model, name="auto", iterations=1)
    assert type(opt).__name__ == "AdamW"
    assert trainer.args.warmup_bias_lr == 0.0
```
image

**Comment by glenn-jocher** [comment]
The logic in your snippet is right; to keep CI fast and avoid any dataset download during test collection, please add it to the PR as a true unit test that patches `DetectionTrainer.get_dataset()` during init (the behavior you’re validating is in [`BaseTrainer.build_optimizer()`](https://docs.ultralytics.com/reference/engine/trainer/)):

```python
from unittest import mock

from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer


def test_auto_optimizer_warmup_bias_lr():
    with mock.patch.object(DetectionTrainer, "get_dataset", return_value={"train": "", "val": "", "nc": 80}):
        trainer = DetectionTrainer(overrides={"model": "yolo26n.yaml"})

    model = YOLO("yolo26n.yaml").model

    trainer.args.warmup_bias_lr = 0.1
    opt = trainer.build_optimizer(model, name="auto", iterations=20001)
    assert type(opt).__name__ == "MuSGD"
    assert trainer.args.warmup_bias_lr == 0.1

    trainer.args.warmup_bias_lr = 0.1
    opt = trainer.build_optimizer(model, name="auto", iterations=1)
    assert type(opt).__name__ == "AdamW"
    assert trainer.args.warmup_bias_lr == 0.0
```

Once that’s committed/pushed to the PR branch (no screenshots needed), we’ll rely on CI to validate it as part of the review.

**Comment by glenn-jocher** [comment]
Thanks for this fix. It's byte-for-byte identical to the sibling PR **#23831**, which targets the same `warmup_bias_lr` MuSGD-vs-AdamW behavior and additionally links the tracking issue (`Fixes #23637`). Since the two conflict and only one can land, I'm consolidating on #23831 and closing this one.

The unit test you drafted in the review thread here is valuable — we'll port it onto #23831 before merging so the `optimizer=auto` → `warmup_bias_lr` behavior is locked in. Thanks for the thorough back-and-forth on this!
